### PR TITLE
Supersede #46

### PR DESCRIPTION
**Pull request by warsaw**
This supersedes #46 by adding everything that wasn't the snapcraft.yaml (added in a separate PR).  It cleans everything up, at least as best I can for now.  It also makes all the tests pass.

### CODE REVIEW

**Comment by warsaw** [comment]
Oh yeah, this also adds the `mkfs.vfat` options to match u-d-f so maybe that will "fix" the corruption (although we don't really know what's causing that).

**Comment by ogra1** [comment]
it wont fix the corruption, i spent a whole day trying all possible option combos and i'm pretty confident after this that it is not related to mkfs ... rather to mtools or to the way we align the partition content with the table once we assemble the final image.

File: ubuntu_image/builder.py [hunk 1]
```diff
@@ -333,6 +340,9 @@ def populate_filesystems(self):
                     os.path.join(part_dir, filename)
                     for filename in os.listdir(part_dir)
                     )
+                # XXX Make sure the $PATH gets propagated to the mcopy(1)
+                # subprocess.  For unknown reasons it might not yet be set
+                # up.  See PR#46 for details.
```
Comment (mvo5) [inline ubuntu_image/builder.py]: Hm, unless I miss something the reason is the following:
```
$ python3 
Python 3.5.2 (default, Jul  5 2016, 12:43:10) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from ubuntu_image.helpers import run
>>> run("env", env=dict(MTOOLS_SKIP_CHECK='1'), stdout=None)
MTOOLS_SKIP_CHECK=1
CompletedProcess(args=['env'], returncode=0, stderr='')
```
I.e. once you set `env` to something it replaces the existing env and does not merge it(compare the output to a run() without env, here all the user env is still there. This means of course that the custom PATH that is set by the snapcraft wrapper will not work. Also the LD_LIBRARY_PATH etc are gone. Fwiw, without inheriting env (e.g. via `run("env", env=dict(MTOOLS_SKIP_CHECK='1').update(os.environ), stdout=None)`) here and on the other places that set `env` the snap won't work.

File: ubuntu_image/builder.py [hunk 2]
```diff
@@ -297,8 +302,10 @@ def prepare_filesystems(self):
             raise AssertionError('No room for root filesystem data')
         self.rootfs_size = avail_space
         self.root_img = os.path.join(self.images, 'root.img')
-        run('dd if=/dev/zero of={} count=0 bs={}M seek=1'.format(
-            self.root_img, avail_space))
+        # Create empty file with holes.
```
Comment (mvo5) [inline ubuntu_image/builder.py]: Suggesting changes to my own writing here :) But maybe `Create sparse file` to use a more matching technical term?

File: ubuntu_image/helpers.py [hunk 3]
```diff
@@ -91,34 +96,35 @@ def inner(*args, **kws):
...
-    # prepare-image` from failing with an unverified signature on the
-    # model.assertion.  We obviously can't sign it with the real root keys,
-    # and there's no other way to inject testing data.
-    #
-    # $PATH is needed because without it `snap prepare-image` can't find
-    # /usr/bin/squashfs.  This is currently unexplained.
-    env = dict(UBUNTU_IMAGE_SKIP_COPY_UNVERIFIED_MODEL='1',
-               PATH='/usr/bin')
-    run(cmd, env=env)
+    run(cmd, stdout=None, stderr=None, env=dict(PATH=os.environ['PATH']))
```
Comment (mvo5) [inline ubuntu_image/helpers.py]: See my comment about env from above, its probably ok now if we just remove `env=...` here.

**Comment by mvo5** [comment]
Nice, thanks for the test fixes!

File: ubuntu_image/image.py [hunk 4]
```diff
@@ -55,17 +55,15 @@ def copy_blob(self, blob_path, **dd_args):
...
+        args = SPACE.join('{}={}'.format(key, value)
+                          for key, value in dd_args.items())
         # Run the command.  We'll capture stderr for logging purposes.
         #
         # TBD:
         # - check status of the returned CompletedProcess
         # - handle errors
         # - log stdout/stderr
-        run(args, stdout=PIPE, stderr=PIPE, universal_newlines=True)
+        run('dd of={} if={} {} conv=sparse'.format(self.path, blob_path, args))
```
Comment (vorlonofportland) [inline ubuntu_image/image.py]: No, this is a regression in support for path names that contain spaces.  We should always prefer arg arrays over a space-split string where possible, especially where we are dealing with arguments including paths that the user can influence.

Will fix up on the branch and push for re-testing.

File: ubuntu_image/image.py [hunk 5]
```diff
@@ -179,8 +176,7 @@ def partition(self, partnum, **sfdisk_args):
         # - check status of the returned CompletedProcess
         # - handle errors
         # - log stdout/stderr
-        run(args, input=input, stdout=PIPE, stderr=PIPE,
-            universal_newlines=True)
+        run(SPACE.join(args))
```
Comment (vorlonofportland) [inline ubuntu_image/image.py]: and this broke MBR-based images, because sfdisk hangs forever waiting for input...

**Pull request merged**